research/context/preselection/preselect_wikipedia_comments.py

```diff
@@ -0,0 +1,125 @@
+    :param high_tox: comments should be lower than that
+    :param limit: fetch some
+    :param type: kind of comments (wikipedia specific)
+    :param max_txt_len: max length of comment
+    :return: dataframe with results
+    """
+    assert low_tox>=0 and high_tox<=1
+    query = """
+    SELECT *, rand() as r 
+        FROM `{gbq_table_name}`
+        WHERE length(cleaned_content)>1 AND length(cleaned_content)<{max_txt_len} AND replyTo_id!="null" AND type="{typ}" AND toxicity>{low} AND toxicity<{high}
+        order by r
+        LIMIT {lim}
+    """.format(gbq_table_name=gbq_table_name, low=low_tox, high=high_tox, lim=limit, typ=type, max_txt_len=max_txt_len, min_txt_len=min_txt_len)
+    query_job = client.query(query)
+    results = query_job.result()
+    return results.to_dataframe()
+
+
+def sample(gbq_table_name, toxic_zones_num, min_txt_len, max_txt_len, type, limit, randomly_selected_limit):
+    """
+    Create a sample of comments using toxic zones and a random batch.
+    :param toxic_zones_num:
+    :return: toxic_zones_num samples + one with randomly selected comments
+    """
+    # Create random sample (initialization)
+    table = query(low_tox=0, high_tox=1, min_txt_len=min_txt_len, max_txt_len=max_txt_len, limit=randomly_selected_limit, type=type, gbq_table_name=gbq_table_name)
+    table["toxic_zone"] = [-1] * table.shape[0]
+    # Add samples following a toxicity-preselection
+    for i in tqdm(range(toxic_zones_num)):
+        preselected = query(low_tox=float(i)/toxic_zones_num, high_tox=float(i+1)/toxic_zones_num, min_txt_len=min_txt_len, max_txt_len=max_txt_len, limit=limit, type=type, gbq_table_name=gbq_table_name)
+        preselected["toxic_zone"] = [i] * preselected.shape[0]
+        # Stack to initialized/updated table
+        table = pd.concat([table, preselected], axis=0)
+    # Shuffle
+    table = table.sample(frac=1).reset_index(drop=True)
+    return table
+
+
+if __name__ == "__main__":
+    assert FLAGS.gbq_project_name is not None
+    assert (FLAGS.output_json_path is None and FLAGS.output_csv_path is not None) or (FLAGS.output_json_path is not None and FLAGS.output_csv_path is None)
+    assert FLAGS.comment_type in {'ADDITION', 'DELETION', 'CREATION', 'MODIFICATION'}
+    # connect to gbq, sample and save
+    client = connect_to_gbq(FLAGS.gbq_project_name)
+    table = sample(gbq_table_name=FLAGS.gbq_table_name, toxic_zones_num=FLAGS.zones_num, min_txt_len=FLAGS.min_txt_len, max_txt_len=FLAGS.max_txt_len, limit=FLAGS.limit, type=FLAGS.comment_type, randomly_selected_limit=FLAGS.randomly_selected_limit)
+    if FLAGS.output_csv_path is not None:
+        # e.g., "sample.11-toxic-zones.addition.csv"
+        table.to_csv(FLAGS.output_csv_path, index=False)
+    elif FLAGS.output_json_path is not None:
+        # e.g., "sample.11-toxic-zones.addition.json"
+        table.to_json(path_or_buf=FLAGS.output_json_path, orient="records", lines=True)
+    else:
+        print ("Not implemented yet.")
```

Comment:
Better to say that either CSV or JSON should be specified. Also better to put this error at the start, so that we don't do the expensive bigquery work only to get an error.

Reply:
I guess that the assertion (2nd row of the function) saves from this expense, but I also add this in the flags info.